test_getaddrinfo failure

I'm running the tests on Alpine Linux with Python 3.8 and all the newest dependencies as of the time of writing. All tests succeed except for `test_getaddrinfo` in `trio/tests/test_socket.py`:

```
============================================================= FAILURES ==============================================================
_________________________________________________________ test_getaddrinfo __________________________________________________________

monkeygai = <trio.tests.test_socket.MonkeypatchedGAI object at 0x7fcd6388ff10>

    async def test_getaddrinfo(monkeygai):
        def check(got, expected):
            # win32 returns 0 for the proto field
            def without_proto(gai_tup):
                return gai_tup[:2] + (0,) + gai_tup[3:]
    
            expected2 = [without_proto(gt) for gt in expected]
            assert got == expected or got == expected2
    
        # Simple non-blocking non-error cases, ipv4 and ipv6:
        with assert_checkpoints():
            res = await tsocket.getaddrinfo(
                "127.0.0.1", "12345", type=tsocket.SOCK_STREAM
            )
    
>       check(res, [
            (tsocket.AF_INET,  # 127.0.0.1 is ipv4
             tsocket.SOCK_STREAM,
             tsocket.IPPROTO_TCP,
             "",
             ("127.0.0.1", 12345)),
        ])  # yapf: disable

trio/tests/test_socket.py:116: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

got = [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '127.0.0.1', ('127.0.0.1', 12345))]
expected = [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 12345))]

    def check(got, expected):
        # win32 returns 0 for the proto field
        def without_proto(gai_tup):
            return gai_tup[:2] + (0,) + gai_tup[3:]
    
        expected2 = [without_proto(gt) for gt in expected]
>       assert got == expected or got == expected2
E       AssertionError: assert ([(<AddressFam...0.1', 12345))] == [(<AddressFam...0.1', 12345))]
E         At index 0 diff: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '127.0.0.1', ('127.0.0.1', 12345)) != (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 12345))
E         Use -v to get the full diff or [(<AddressFam...0.1', 12345))] == [(<AddressFam...0.1', 12345))]
E         At index 0 diff: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '127.0.0.1', ('127.0.0.1', 12345)) != (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0, '', ('127.0.0.1', 12345))
E         Use -v to get the full diff)

trio/tests/test_socket.py:108: AssertionError
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -43,8 +43,8 @@ jobs:
           # Should match 'name:' up above
           JOB_NAME: 'Windows (${{ matrix.python }}, ${{ matrix.arch }}${{ matrix.extra_name }})'
 
-  Linux:
-    name: 'Linux (${{ matrix.python }}${{ matrix.extra_name }})'
+  Ubuntu:
+    name: 'Ubuntu (${{ matrix.python }}${{ matrix.extra_name }})'
     timeout-minutes: 10
     runs-on: 'ubuntu-latest'
     strategy:
@@ -74,7 +74,7 @@ jobs:
           CHECK_DOCS: '${{ matrix.check_docs }}'
           CHECK_FORMATTING: '${{ matrix.check_formatting }}'
           # Should match 'name:' up above
-          JOB_NAME: 'Linux (${{ matrix.python }}${{ matrix.extra_name }})'
+          JOB_NAME: 'Ubuntu (${{ matrix.python }}${{ matrix.extra_name }})'
 
   macOS:
     name: 'macOS (${{ matrix.python }})'
@@ -96,3 +96,21 @@ jobs:
         env:
           # Should match 'name:' up above
           JOB_NAME: 'macOS (${{ matrix.python }})'
+
+  Alpine:
+    name: 'Alpine'
+    timeout-minutes: 10
+    runs-on: 'ubuntu-latest'
+    container: 'alpine'
+    strategy:
+      fail-fast: false
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Install bash
+        run: apk add --no-cache bash
+      - name: Run tests
+        run: ./ci.sh
+        env:
+          # Should match 'name:' up above
+          JOB_NAME: 'Alpine'

--- a/ci.sh
+++ b/ci.sh
@@ -34,43 +34,12 @@ function curl-harder() {
 # Bootstrap python environment, if necessary
 ################################################################
 
-### Azure pipelines + Windows ###
+### Alpine ###
 
-# On azure pipeline's windows VMs, to get reasonable performance, we need to
-# jump through hoops to avoid touching the C:\ drive as much as possible.
-if [ "$AGENT_OS" = "Windows_NT" ]; then
-    # By default temp and cache directories are on C:\. Fix that.
-    export TEMP="${AGENT_TEMPDIRECTORY}"
-    export TMP="${AGENT_TEMPDIRECTORY}"
-    export TMPDIR="${AGENT_TEMPDIRECTORY}"
-    export PIP_CACHE_DIR="${AGENT_TEMPDIRECTORY}\\pip-cache"
-
-    # Download and install Python from scratch onto D:\, instead of using the
-    # pre-installed versions that azure pipelines provides on C:\.
-    # Also use -DirectDownload to stop nuget from caching things on C:\.
-    nuget install "${PYTHON_PKG}" -Version "${PYTHON_VERSION}" \
-          -OutputDirectory "$PWD/pyinstall" -ExcludeVersion \
-          -Source "https://api.nuget.org/v3/index.json" \
-          -Verbosity detailed -DirectDownload -NonInteractive
-
-    pydir="$PWD/pyinstall/${PYTHON_PKG}"
-    export PATH="${pydir}/tools:${pydir}/tools/scripts:$PATH"
-fi
-
-### Travis + macOS ###
-
-if [ "$TRAVIS_OS_NAME" = "osx" ]; then
-    JOB_NAME="osx_${MACPYTHON}"
-    curl-harder -o macpython.pkg https://www.python.org/ftp/python/${MACPYTHON}/python-${MACPYTHON}-macosx10.6.pkg
-    sudo installer -pkg macpython.pkg -target /
-    ls /Library/Frameworks/Python.framework/Versions/*/bin/
-    PYTHON_EXE=/Library/Frameworks/Python.framework/Versions/*/bin/python3
-    # The pip in older MacPython releases doesn't support a new enough TLS
-    curl-harder -o get-pip.py https://bootstrap.pypa.io/get-pip.py
-    sudo $PYTHON_EXE get-pip.py
-    sudo $PYTHON_EXE -m pip install virtualenv
-    $PYTHON_EXE -m virtualenv testenv
-    source testenv/bin/activate
+if [ -e /etc/alpine-release ]; then
+    apk add --no-cache gcc musl-dev libffi-dev openssl-dev python3-dev curl
+    python3 -m venv venv
+    source venv/bin/activate
 fi
 
 ### PyPy nightly (currently on Travis) ###

--- a/newsfragments/1499.bugfix.rst
+++ b/newsfragments/1499.bugfix.rst
@@ -1,0 +1,2 @@
+Fixed an over-strict test that caused failures on Alpine Linux.
+Started testing against Alpine in CI.

--- a/trio/tests/test_socket.py
+++ b/trio/tests/test_socket.py
@@ -151,8 +151,10 @@ async def test_getaddrinfo(monkeygai):
     with assert_checkpoints():
         with pytest.raises(tsocket.gaierror) as excinfo:
             await tsocket.getaddrinfo("::1", "12345", type=-1)
-    # Linux, Windows
+    # Linux + glibc, Windows
     expected_errnos = {tsocket.EAI_SOCKTYPE}
+    # Linux + musl
+    expected_errnos.add(tsocket.EAI_SERVICE)
     # macOS
     if hasattr(tsocket, "EAI_BADHINTS"):
         expected_errnos.add(tsocket.EAI_BADHINTS)

--- a/trio/tests/test_socket.py
+++ b/trio/tests/test_socket.py
@@ -101,11 +101,19 @@ def test_socket_has_some_reexports():
 async def test_getaddrinfo(monkeygai):
     def check(got, expected):
         # win32 returns 0 for the proto field
-        def without_proto(gai_tup):
-            return gai_tup[:2] + (0,) + gai_tup[3:]
+        # alpine and glibc have inconsistent handling of the canonical name
+        # field (https://github.com/python-trio/trio/issues/1499)
+        # Neither gets used much and there isn't much opportunity for us to
+        # mess them up, so we don't bother checking them
+        def interesting_fields(gai_tup):
+            # (family, type, proto, canonname, sockaddr)
+            family, type, proto, canonname, sockaddr = gai_tup
+            return (family, type, sockaddr)
 
-        expected2 = [without_proto(gt) for gt in expected]
-        assert got == expected or got == expected2
+        def filtered(gai_list):
+            return [interesting_fields(gai_tup) for gai_tup in gai_list]
+
+        assert filtered(got) == filtered(expected)
 
     # Simple non-blocking non-error cases, ipv4 and ipv6:
     with assert_checkpoints():

--- a/trio/tests/test_socket.py
+++ b/trio/tests/test_socket.py
@@ -101,10 +101,10 @@ def test_socket_has_some_reexports():
 async def test_getaddrinfo(monkeygai):
     def check(got, expected):
         # win32 returns 0 for the proto field
-        # alpine and glibc have inconsistent handling of the canonical name
+        # musl and glibc have inconsistent handling of the canonical name
         # field (https://github.com/python-trio/trio/issues/1499)
-        # Neither gets used much and there isn't much opportunity for us to
-        # mess them up, so we don't bother checking them
+        # Neither field gets used much and there isn't much opportunity for us
+        # to mess them up, so we don't bother checking them here
         def interesting_fields(gai_tup):
             # (family, type, proto, canonname, sockaddr)
             family, type, proto, canonname, sockaddr = gai_tup

